src/datasets/io/sql.py

```diff
@@ -0,0 +1,136 @@
+import contextlib
+import multiprocessing
+import os
+from sqlite3 import Connection, connect
+from typing import Optional, Union
+
+from .. import Dataset, Features, config
+from ..formatting import query_table
+from ..packaged_modules.sql.sql import Sql
+from ..utils import logging
+from ..utils.typing import NestedDataStructureLike, PathLike
+from .abc import AbstractDatasetInputStream
+
+
+class SqlDatasetReader(AbstractDatasetInputStream):
+    def __init__(
+        self,
+        conn: NestedDataStructureLike[Union[PathLike, Connection]],
+        table_name: str,
+        features: Optional[Features] = None,
+        cache_dir: str = None,
+        keep_in_memory: bool = False,
+        **kwargs,
+    ):
+        super().__init__(features=features, cache_dir=cache_dir, keep_in_memory=keep_in_memory, **kwargs)
+        conn = conn if isinstance(conn, dict) else {"train": conn}
+        self.builder = Sql(
+            cache_dir=cache_dir,
+            conn=conn,
+            features=features,
+            table_name=table_name,
+            **kwargs,
+        )
+
+    def read(self):
+        download_config = None
+        download_mode = None
+        ignore_verifications = False
+        use_auth_token = None
+        base_path = None
+
+        self.builder.download_and_prepare(
+            download_config=download_config,
+            download_mode=download_mode,
+            ignore_verifications=ignore_verifications,
+            # try_from_hf_gcs=try_from_hf_gcs,
+            base_path=base_path,
+            use_auth_token=use_auth_token,
+        )
+
+        # Build dataset for splits
+        dataset = self.builder.as_dataset(
+            split="train", ignore_verifications=ignore_verifications, in_memory=self.keep_in_memory
```

Comment:
Same as `GeneratorDatasetInputStream` where only the train set is possible. let me know if that's alright.